hypothesis-python/src/hypothesis/extra/gufunc.py

```diff
@@ -0,0 +1,475 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Most of this work is copyright (C) 2013-2019 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+# This module uses the numpy parser of the Generalized Universal Function API
+# signatures `_parse_gufunc_signature`, which is only available in
+# numpy>=1.12.0 and therefore requires a bump in the requirements for
+# hypothesis.
+from __future__ import absolute_import, division, print_function
+
+import string
+from collections import defaultdict
+
+import numpy as np
+import numpy.lib.function_base as npfb
+
+from hypothesis.errors import InvalidArgument
+from hypothesis.extra.numpy import arrays, order_check
+from hypothesis.internal.compat import integer_types
+from hypothesis.internal.validation import check_type, check_valid_bound
+from hypothesis.searchstrategy import SearchStrategy
+from hypothesis.strategies import (
+    builds,
+    composite,
+    fixed_dictionaries,
+    integers,
+    just,
+    tuples,
+)
+
+__all__ = ["gufunc_args", "gufunc_arg_shapes"]
+
+# Should not ever need to broadcast beyond this, but should be able to set it
+# as high as 32 before breaking assumptions in numpy.
+GLOBAL_DIMS_MAX = 12
+
+
+# Key used in min_side and max_side to indicate min/max on broadcasted dims,
+# building sentinel class so we have clean __repr__.
+class _BcastDimType(object):
+    def __repr__(self):
+        return "BCAST_DIM"
+
+
+BCAST_DIM = _BcastDimType()
+
+# Value used in default dict for max side if variable not specified
+DEFAULT_MAX_SIDE = 5
+
+# This uses "private" function of numpy, but it does the job. It throws a
+# pretty readable exception for invalid input, we don't need to add anything.
+parse_gufunc_signature = npfb._parse_gufunc_signature
```

Comment:
I would suggest inlining this to it's single use.  We should detect and give a useful error message if Numpy is too old to support this feature.

Reply:
By that you mean putting the line inside the `gufunc_arg_shapes` function? I have it on the outside so we can test its behavior is as expected.  Since it is a private function of numpy it is probably a good idea to have a test for it.

Are you suggesting a try-except in case numpy takes it out?

Reply:
Yep, putting it inside the function - and behind a guard like 

```python
if not hasattr(npfb, "_parse_gufunc_signature"):
    raise HypothesisException("This function requires Numpy 1.12 or later.")
```

IMO running all the gufunc tests are sufficient, we don't need to test the Numpy implementation separately.  However we might want to think about adding a test job against `numpy==1.12.0` in addition to the current latest numpy tests.